crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -152,57 +161,308 @@ impl RawModuleResolutionSettings {
             .into_iter()
-            .map(|fs_path| ModuleResolutionPathBuf::extra(fs_path).unwrap())
+            .map(|fs_path| {
+                Arc::new(
+                    ModuleResolutionPathBuf::extra(SystemPath::absolute(
+                        fs_path,
+                        current_directory,
+                    ))
+                    .unwrap(),
+                )
+            })
             .collect();
 
-        paths.push(ModuleResolutionPathBuf::first_party(workspace_root).unwrap());
-
-        paths.push(
-            custom_typeshed.map_or_else(ModuleResolutionPathBuf::vendored_stdlib, |custom| {
-                ModuleResolutionPathBuf::stdlib_from_custom_typeshed_root(&custom).unwrap()
-            }),
+        let workspace_root = Arc::new(
+            ModuleResolutionPathBuf::first_party(SystemPath::absolute(
+                workspace_root,
+                current_directory,
+            ))
+            .unwrap(),
         );
 
+        let stdlib = Arc::new(custom_typeshed.map_or_else(
+            ModuleResolutionPathBuf::vendored_stdlib,
+            |custom| {
+                ModuleResolutionPathBuf::stdlib_from_custom_typeshed_root(&SystemPath::absolute(
+                    custom,
+                    current_directory,
+                ))
+                .unwrap()
+            },
+        ));
+
         // TODO vendor typeshed's third-party stubs as well as the stdlib and fallback to them as a final step
-        if let Some(site_packages) = site_packages {
-            paths.push(ModuleResolutionPathBuf::site_packages(site_packages).unwrap());
-        }
+        let site_packages = site_packages.map(|path| {
+            Arc::new(
+                ModuleResolutionPathBuf::site_packages(SystemPath::absolute(
+                    path,
+                    current_directory,
+                ))
+                .unwrap(),
+            )
+        });
 
         ModuleResolutionSettings {
             target_version,
-            search_paths: OrderedSearchPaths(paths.into_iter().map(Arc::new).collect()),
+            search_path_settings: ValidatedSearchPathSettings {
+                extra_paths,
+                workspace_root,
+                stdlib,
+                site_packages,
+            },
         }
     }
 }
 
-/// A resolved module resolution order as per the [typing spec]
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct ValidatedSearchPathSettings {
+    extra_paths: Vec<SearchPathRoot>,
+    workspace_root: SearchPathRoot,
+    stdlib: SearchPathRoot,
+    site_packages: Option<SearchPathRoot>,
+}
+
+impl ValidatedSearchPathSettings {
+    /// Implementation of the typing spec's [module resolution order].
+    ///
+    /// It's important that we collect these instead of iterating over them lazily,
+    /// so that we can be sure that it doesn't contain any duplicates.
+    /// At runtime, [`sys.path` never contains duplicate entries]
```

Comment:
Considering that we build this set (this is the validated settings). Can't we remove duplicates when constructing this struct. It seems wasteful to repeat the deduplication on every iterator call.

Reply:
Yes, I wondered about this. I think it makes it slightly less clear to the reader exactly what data is being stored where, but I agree it would be more efficient. I'll make the change.